src/llm/io_processing/output_parser.cpp

```diff
@@ -120,55 +202,110 @@ static inline bool chunkContainsSpecialParsingTag(const std::string& chunk, cons
 
 std::optional<rapidjson::Document> OutputParser::parseChunk(const std::string& chunkResponse, const bool toolsAvailable, ov::genai::GenerationFinishReason finishReason) {
     // Using appropriate parser based on the current processing phase
-    // Call to this method should always return either result from parser parseChunk implementation or common parseContentChunk method.
-    // If for any processing phase a nullopt should be returned, it should be done in the parser implementation.
-    // Do not return nullopt directly from this method.
+    // Call to this method should return either result from parser parseChunk implementation when we are in particular phase
+    // or std::nullopt when we are waiting for more chunks to determine if we should switch phase or not.
 
     bool reasoningParserExistsAndSupportsStreaming = reasoningParser && !reasoningParser->getParsingStartTag().empty() && !reasoningParser->getParsingEndTag().empty();
     bool toolParserExistsAndSupportsStreaming = toolParser && !toolParser->getParsingStartTag().empty();
     bool applyToolParser = toolParserExistsAndSupportsStreaming && toolsAvailable;
 
+    streamOutputCache.add(chunkResponse);
+
     if (processingPhase == UNKNOWN) {
         // If we are in the UNKNOWN phase, we need to determine if we should switch to CONTENT, REASONING, or TOOL_CALLS phase.
-        if (reasoningParserExistsAndSupportsStreaming && isParsingTagPartOfChunk(chunkResponse, reasoningParser->getParsingStartTag())) {
-            processingPhase = REASONING;
-            return reasoningParser->parseChunk(chunkResponse, finishReason);
-        } else if (applyToolParser) {
-            if (isParsingTagPartOfChunk(chunkResponse, toolParser->getParsingStartTag()) || chunkContainsSpecialParsingTag(chunkResponse, toolParser->getSpecialParsingStartTags())) {
-                processingPhase = TOOL_CALLS;
-                return toolParser->parseChunk(chunkResponse, finishReason);
-            } else if (toolParser->isImmediateParsingEnabled()) {
+        TagLookupStatus anyStartTagStatus = TagLookupStatus::NOT_FOUND;
+        if (reasoningParserExistsAndSupportsStreaming) {
+            // Check if reasoning start tag has been received
+            TagLookupStatus reasoningStartTagStatus = streamOutputCache.lookupTag(reasoningParser->getParsingStartTag());
+            if (reasoningStartTagStatus == TagLookupStatus::NOT_FOUND) {
+                // If reasoning start tag is not found, check if any of the special start tags are found
+                reasoningStartTagStatus = streamOutputCache.lookupTags(reasoningParser->getSpecialParsingStartTags());
+            }
+            if (reasoningStartTagStatus == TagLookupStatus::FOUND_COMPLETE) {
+                auto result = reasoningParser->parseChunk(streamOutputCache.getBuffer(), finishReason);
+                processingPhase = REASONING;
+                streamOutputCache.clear();
+                return result;
```

Comment:
We should clean cache before every non-nullopt return, so to make it less error prone we could try to tie it to the core logic (like parsing itself)